firewood/src/merkle.rs

```diff
@@ -1989,4 +1962,216 @@ mod tests {
+
+        assert_eq!(
+            merkle.get(&key, root).unwrap().as_deref(),
+            Some(val.as_slice())
+        );
+
+        assert_eq!(
+            merkle.get(&key_2, root).unwrap().as_deref(),
+            Some(val_2.as_slice())
+        );
+
+        assert_eq!(
+            merkle.get(&key_3, root).unwrap().as_deref(),
+            Some(val_3.as_slice())
+        );
+    }
+
+    #[test]
+    fn old_branch_is_a_child_of_new_branch() {
+        let key = vec![0xff, 0xf0];
+        let val = vec![1];
+        let key_2 = vec![0xff, 0xf0, 0x00];
+        let val_2 = vec![2];
+        let key_3 = vec![0xff];
+        let val_3 = vec![3];
+
+        let mut merkle = create_test_merkle();
+        let root = merkle.init_root().unwrap();
+
+        merkle.insert(&key, val.clone(), root).unwrap();
+        merkle.insert(&key_2, val_2.clone(), root).unwrap();
+        merkle.insert(&key_3, val_3.clone(), root).unwrap();
+
+        assert_eq!(
+            merkle.get(&key, root).unwrap().as_deref(),
+            Some(val.as_slice())
+        );
+
+        assert_eq!(
+            merkle.get(&key_2, root).unwrap().as_deref(),
+            Some(val_2.as_slice())
+        );
+
+        assert_eq!(
+            merkle.get(&key_3, root).unwrap().as_deref(),
+            Some(val_3.as_slice())
+        );
+    }
+
+    #[test]
+    fn overlapping_branch_insert() {
+        let key = vec![0xff];
+        let val = vec![1];
+        let key_2 = vec![0xff, 0x00];
+        let val_2 = vec![2];
+
+        let overwrite = vec![3];
+
+        let mut merkle = create_test_merkle();
+        let root = merkle.init_root().unwrap();
+
+        merkle.insert(&key, val.clone(), root).unwrap();
+        merkle.insert(&key_2, val_2.clone(), root).unwrap();
+
+        assert_eq!(
+            merkle.get(&key, root).unwrap().as_deref(),
+            Some(val.as_slice())
+        );
+
+        assert_eq!(
+            merkle.get(&key_2, root).unwrap().as_deref(),
+            Some(val_2.as_slice())
+        );
+
+        merkle.insert(&key, overwrite.clone(), root).unwrap();
+
+        assert_eq!(
+            merkle.get(&key, root).unwrap().as_deref(),
+            Some(overwrite.as_slice())
+        );
+
+        assert_eq!(
+            merkle.get(&key_2, root).unwrap().as_deref(),
+            Some(val_2.as_slice())
+        );
+    }
+
+    #[test]
+    fn single_key_proof_with_one_node() {
+        let mut merkle = create_test_merkle();
+        let root = merkle.init_root().unwrap();
+        let key = b"key";
+        let value = b"value";
+
+        merkle.insert(key, value.to_vec(), root).unwrap();
+        let root_hash = merkle.root_hash(root).unwrap();
+
+        let proof = merkle.prove(key, root).unwrap();
+
+        let verified = proof.verify(key, root_hash.0).unwrap();
+
+        assert_eq!(verified, Some(value.to_vec()));
+    }
+
+    #[test]
+    fn two_key_proof_without_shared_path() {
```

Comment:
Does this cover something not covered in `test_proof`?

Reply:
I don't know and it's impossible to tell what the structure of the tree is in that test without debugging. There is no guarantee that there are nodes that don't share a path.

Reply:
It would be easy to add a check for that in the other test and fold this one into that. It does generate enough data that it seems extremely likely there will be non-shared path proofs.

I like the idea of testing more than just this one case.

Reply:
But that goes against proper unit testing principles. Unit tests should just test one thing. I'm not dogmatic about this; I've written tests that do more, but I had an absolute nightmare trying to figure out why `test_proof` was failing before. With this case, it would be ultra clear why it's failing; there's no extra complexity. 

We also had another test that was only passing because the leaf nodes that were created didn't have partial paths BY CHANCE. We should know the exact state of the thing we are testing. We should *also* fuzz test (using proper fuzzer, not just random values), but fuzzing and unit testing should be completely separate IMO. They serve a different purpose.

Reply:
Fair point. We should rename that test fuzz or something.

Reply:
I do wonder if there are other cases besides this one we aren't covering though, but I can't offhand think of any.